In [1]:
import pyspark
from pyspark.sql import SparkSession
from pyspark.sql import types

### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?
- **3.3.2**
- 2.1.4
- 1.2.3
- 5.4
</br></br>



In [2]:
spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

23/03/12 17:14:36 WARN Utils: Your hostname, Vostro-14-5401 resolves to a loopback address: 127.0.1.1; using 192.168.1.115 instead (on interface wlp1s0)
23/03/12 17:14:36 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


23/03/12 17:14:36 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
23/03/12 17:14:37 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
23/03/12 17:14:37 WARN Utils: Service 'SparkUI' could not bind on port 4041. Attempting port 4042.
23/03/12 17:14:37 WARN Utils: Service 'SparkUI' could not bind on port 4042. Attempting port 4043.
23/03/12 17:14:37 WARN Utils: Service 'SparkUI' could not bind on port 4043. Attempting port 4044.
23/03/12 17:14:37 WARN Utils: Service 'SparkUI' could not bind on port 4044. Attempting port 4045.


In [3]:
spark.version

'3.3.2'


### Question 2: 

**HVFHW June 2021**

Read it with Spark using the same schema as we did in the lessons.</br> 
We will use this dataset for all the remaining questions.</br>
Repartition it to 12 partitions and save it to parquet.</br>
What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.</br>


- 2MB
- **24MB**
- 100MB
- 250MB
</br></br>


In [5]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz -P ./data/raw/fhvhv/
    

--2023-03-12 17:19:10--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz
Распознаётся github.com (github.com)… 140.82.121.3
Подключение к github.com (github.com)|140.82.121.3|:443... соединение установлено.
HTTP-запрос отправлен. Ожидание ответа… 302 Found
Адрес: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e-a6da-4923-ad6f-35ff02446a51?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230312%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230312T141910Z&X-Amz-Expires=300&X-Amz-Signature=502b7d1638b2470b56f0c1e39dfa684fcf8d540dd73bc1c75577976bec4ff52f&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhvhv_tripdata_2021-06.csv.gz&response-content-type=application%2Foctet-stream [переход]
--2023-03-12 17:19:10--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948

In [8]:
!gzip -d ./data/raw/fhvhv/fhvhv_tripdata_2021-06.csv.gz

In [9]:
!ls -lh ./data/raw/fhvhv/fhvhv_tripdata_2021-06.csv

-rw-rw-r-- 1 tikhonov tikhonov 878M дек 20 03:13 ./data/raw/fhvhv/fhvhv_tripdata_2021-06.csv


In [68]:
schema = types.StructType([
    types.StructField('dispatching_base_num', types.StringType(), True),
    types.StructField('pickup_datetime', types.TimestampType(), True),
    types.StructField('dropoff_datetime', types.TimestampType(), True),
    types.StructField('PULocationID', types.IntegerType(), True),
    types.StructField('DOLocationID', types.IntegerType(), True),
    types.StructField('SR_Flag', types.StringType(), True)
])

In [69]:
df = spark.read \
    .option("header", "true") \
    .schema(schema) \
    .csv('data/raw/fhvhv/fhvhv_tripdata_2021-06.csv')

df = df.repartition(12)

df.write.parquet('data/pq/fhvhv/2021/06/',mode='overwrite')

23/03/12 20:35:13 WARN CSVHeaderChecker: Number of column in CSV header is not equal to number of fields in the schema:
 Header length: 7, schema size: 6
CSV file: file:///opt/Python/de_zoocamp/week_5/oficial_repo/code/data/raw/fhvhv/fhvhv_tripdata_2021-06.csv


23/03/12 20:35:58 WARN MemoryManager: Total allocation exceeds 95.00% (1,020,054,720 bytes) of heap memory
Scaling row group sizes to 95.00% for 8 writers


23/03/12 20:36:04 WARN MemoryManager: Total allocation exceeds 95.00% (1,020,054,720 bytes) of heap memory
Scaling row group sizes to 95.00% for 8 writers


In [70]:
!ls -lah data/pq/fhvhv/2021/06

итого 262M
drwxr-xr-x 2 tikhonov tikhonov 4,0K мар 12 20:36 .
drwxr-xr-x 3 tikhonov tikhonov 4,0K мар 12 20:35 ..
-rw-r--r-- 1 tikhonov tikhonov  22M мар 12 20:36 part-00000-5f29ac06-05c1-43a6-960a-a89169b72afa-c000.snappy.parquet
-rw-r--r-- 1 tikhonov tikhonov 173K мар 12 20:36 .part-00000-5f29ac06-05c1-43a6-960a-a89169b72afa-c000.snappy.parquet.crc
-rw-r--r-- 1 tikhonov tikhonov  22M мар 12 20:36 part-00001-5f29ac06-05c1-43a6-960a-a89169b72afa-c000.snappy.parquet
-rw-r--r-- 1 tikhonov tikhonov 173K мар 12 20:36 .part-00001-5f29ac06-05c1-43a6-960a-a89169b72afa-c000.snappy.parquet.crc
-rw-r--r-- 1 tikhonov tikhonov  22M мар 12 20:36 part-00002-5f29ac06-05c1-43a6-960a-a89169b72afa-c000.snappy.parquet
-rw-r--r-- 1 tikhonov tikhonov 173K мар 12 20:36 .part-00002-5f29ac06-05c1-43a6-960a-a89169b72afa-c000.snappy.parquet.crc
-rw-r--r-- 1 tikhonov tikhonov  22M мар 12 20:36 part-00003-5f29ac06-05c1-43a6-960a-a89169b72afa-c000.snappy.parquet
-rw-r--r-- 1 tikhonov tikhonov 173K мар 12 20:36 .pa

### Question 3: 

**Count records**  

How many taxi trips were there on June 15?</br></br>
Consider only trips that started on June 15.</br>

- 308,164
- 12,856
- **452,470**
- 50,982
</br></br>


In [71]:
df = spark.read.parquet('data/pq/fhvhv/2021/06/')

In [72]:
from pyspark.sql import functions as F

In [73]:
df \
    .withColumn('pickup_date', F.to_date(df.pickup_datetime)) \
    .filter("pickup_date = '2021-06-15'") \
    .count()

452470

In [74]:
df.createOrReplaceTempView('fhvhv_2021_06')

In [75]:
spark.sql("""
SELECT
    COUNT(1)
FROM 
    fhvhv_2021_06
WHERE
    to_date(pickup_datetime) = '2021-06-15';
""").show()

+--------+
|count(1)|
+--------+
|  452470|
+--------+





### Question 4: 

**Longest trip for each day**  

Now calculate the duration for each trip.</br>
How long was the longest trip in Hours?</br>

- **66.87 Hours**
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours
</br></br>

In [76]:
df.show()

+--------------------+-------------------+-------------------+------------+------------+-------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|
+--------------------+-------------------+-------------------+------------+------------+-------+
|              B02878|2021-06-02 17:32:38|2021-06-02 18:16:03|         237|         144|      N|
|              B02765|2021-06-01 07:17:49|2021-06-01 07:47:22|         216|         139|      N|
|              B02871|2021-06-02 22:05:37|2021-06-02 22:09:33|         256|         255|      N|
|              B02872|2021-06-02 08:02:35|2021-06-02 08:09:48|          56|          82|      N|
|              B02617|2021-06-02 15:49:50|2021-06-02 16:10:47|         100|         262|      N|
|              B02878|2021-06-03 07:38:50|2021-06-03 08:12:16|         180|           7|      N|
|              B02765|2021-06-01 01:35:09|2021-06-01 01:38:38|          63|         258|      N|
|              B02510|2021-06-

In [79]:
query_result = spark.sql("""
SELECT 
pickup_datetime, 
dropoff_datetime,
((bigint(to_timestamp(dropoff_datetime)))-(bigint(to_timestamp(pickup_datetime))))/3600 AS diff_in_hours
FROM fhvhv_2021_06
ORDER BY 3 DESC
LIMIT 10
""")

query_result.show()

+-------------------+-------------------+------------------+
|    pickup_datetime|   dropoff_datetime|     diff_in_hours|
+-------------------+-------------------+------------------+
|2021-06-25 13:55:41|2021-06-28 08:48:25|  66.8788888888889|
|2021-06-22 12:09:45|2021-06-23 13:42:44|25.549722222222222|
|2021-06-27 10:32:29|2021-06-28 06:31:20|19.980833333333333|
|2021-06-26 22:37:11|2021-06-27 16:49:01|18.197222222222223|
|2021-06-23 20:40:43|2021-06-24 13:08:44|16.466944444444444|
|2021-06-23 22:03:31|2021-06-24 12:19:39|14.268888888888888|
|2021-06-24 23:11:00|2021-06-25 13:05:35|13.909722222222221|
|2021-06-04 20:56:02|2021-06-05 08:36:14|             11.67|
|2021-06-27 07:45:19|2021-06-27 19:07:16|11.365833333333333|
|2021-06-20 17:05:12|2021-06-21 04:04:16|10.984444444444444|
+-------------------+-------------------+------------------+



### Question 5: 

**User Interface**

 Spark’s User Interface which shows application's dashboard runs on which local port?</br>

- 80
- 443
- **4040**
- 8080
</br></br>

### Question 6: 

**Most frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)</br>

Using the zone lookup data and the fhvhv June 2021 data, what is the name of the most frequent pickup location zone?</br>

- East Chelsea
- Astoria
- Union Sq
- **Crown Heights North**
</br></br>

**Q6**: Most common locations pair

In [82]:
df_zones = spark.read.parquet('data/pq/zones')

In [83]:
df_zones.columns

['LocationID', 'Borough', 'Zone', 'service_zone']

In [84]:
df.columns

['dispatching_base_num',
 'pickup_datetime',
 'dropoff_datetime',
 'PULocationID',
 'DOLocationID',
 'SR_Flag']

In [86]:
df_zones.createOrReplaceTempView('zones')

In [92]:
df_join = df.join(df_zones, (df['PULocationID'] == df_zones['LocationID']))

df_join.createOrReplaceTempView("trips_joined")

query_result = spark.sql("""
SELECT Zone, COUNT(*)
FROM trips_joined
GROUP BY 1
ORDER BY 2 DESC
LIMIT 10
""")

query_result.show()

+--------------------+--------+
|                Zone|count(1)|
+--------------------+--------+
| Crown Heights North|  231279|
|        East Village|  221244|
|         JFK Airport|  188867|
|      Bushwick South|  187929|
|       East New York|  186780|
|TriBeCa/Civic Center|  164344|
|   LaGuardia Airport|  161596|
|            Union Sq|  158937|
|        West Village|  154698|
|             Astoria|  152493|
+--------------------+--------+

